File: contests/Aoc/2025/7/solution.py
import argparse
from pathlib import Path


def read_grid(path: str) -> list[str]:
    return Path(path).read_text().splitlines()


def find_start(grid: list[str]) -> tuple[int, int] | None:
    for r, row in enumerate(grid):
        c = row.find("S")
        if c != -1:
            return r, c
    return None


def count_splits(grid: list[str]) -> int:
    if not grid:
        return 0

    start = find_start(grid)
    if start is None:
        return 0

    srow, scol = start
    beams = {scol}
    splits = 0

    for r in range(srow + 1, len(grid)):
        next_beams: set[int] = set()
        row = grid[r]
        width = len(row)

        for c in beams:
            if c >= width:
                continue
            cell = row[c]
            if cell == "^":
                splits += 1
                if c > 0:
                    next_beams.add(c - 1)
                if c + 1 < width:
                    next_beams.add(c + 1)
                continue
            next_beams.add(c)

        beams = next_beams
        if not beams:
            break

    return splits


def count_timelines(grid: list[str]) -> int:
    if not grid:
        return 0

    start = find_start(grid)
    if start is None:
        return 0

    srow, scol = start
    beams: dict[int, int] = {scol: 1}

    for r in range(srow + 1, len(grid)):
        next_beams: dict[int, int] = {}
        row = grid[r]
        width = len(row)

        for c, cnt in beams.items():
            if c >= width:
                continue
            if row[c] == "^":
                for nc in (c - 1, c + 1):
                    if 0 <= nc < width:
                        next_beams[nc] = next_beams.get(nc, 0) + cnt
                continue
            next_beams[c] = next_beams.get(c, 0) + cnt

        beams = next_beams
        if not beams:
            break

    return sum(beams.values())


def main() -> None:
    parser = argparse.ArgumentParser(description="Tachyon manifold simulator")
    parser.add_argument("input", nargs="?", default="input.txt", help="path to input file")
    parser.add_argument(
        "--part2", action="store_true", help="compute timeline count (part 2)"
    )
    args = parser.parse_args()

    grid = read_grid(args.input)
    if args.part2:
        print(count_timelines(grid))
    else:
        print(count_splits(grid))


if __name__ == "__main__":
    main()

File: contests/Aoc/2025/7/solution.go
package main

import (
	"bufio"
	"flag"
	"fmt"
	"log"
	"math/big"
	"os"
	"strings"
)

// countSplits simulates the downward beam and returns how many times it hits a splitter.
func countSplits(grid []string) int {
	if len(grid) == 0 {
		return 0
	}

	width := len(grid[0])
	startRow, startCol, found := 0, 0, false
	for r, row := range grid {
		if c := strings.IndexByte(row, 'S'); c != -1 {
			startRow, startCol, found = r, c, true
			break
		}
	}
	if !found {
		return 0
	}

	beams := map[int]struct{}{startCol: {}}
	splits := 0

	for r := startRow + 1; r < len(grid); r++ {
		next := make(map[int]struct{})
		row := grid[r]

		for c := range beams {
			cell := row[c]
			if cell == '^' {
				splits++
				if c > 0 {
					next[c-1] = struct{}{}
				}
				if c+1 < width {
					next[c+1] = struct{}{}
				}
				continue
			}
			next[c] = struct{}{}
		}

		beams = next
		if len(beams) == 0 {
			break // all beams have exited the grid
		}
	}

	return splits
}

// countTimelines returns how many distinct timelines exist after the particle traverses the grid.
func countTimelines(grid []string) *big.Int {
	if len(grid) == 0 {
		return big.NewInt(0)
	}

	width := len(grid[0])
	startRow, startCol, found := 0, 0, false
	for r, row := range grid {
		if c := strings.IndexByte(row, 'S'); c != -1 {
			startRow, startCol, found = r, c, true
			break
		}
	}
	if !found {
		return big.NewInt(0)
	}

	beams := map[int]*big.Int{startCol: big.NewInt(1)}

	for r := startRow + 1; r < len(grid); r++ {
		next := make(map[int]*big.Int)
		row := grid[r]

		for c, count := range beams {
			cell := row[c]
			if cell == '^' {
				if c > 0 {
					addCount(next, c-1, count)
				}
				if c+1 < width {
					addCount(next, c+1, count)
				}
				continue
			}
			addCount(next, c, count)
		}

		beams = next
		if len(beams) == 0 {
			break
		}
	}

	total := big.NewInt(0)
	for _, count := range beams {
		total.Add(total, count)
	}
	return total
}

func addCount(m map[int]*big.Int, col int, add *big.Int) {
	if col < 0 {
		return
	}
	if existing := m[col]; existing != nil {
		existing.Add(existing, add)
		return
	}
	m[col] = new(big.Int).Set(add)
}

func readLines(path string) ([]string, error) {
	file, err := os.Open(path)
	if err != nil {
		return nil, err
	}
	defer file.Close()

	var lines []string
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		lines = append(lines, scanner.Text())
	}

	return lines, scanner.Err()
}

func main() {
	inputPath := "input.txt"
	flag.StringVar(&inputPath, "input", inputPath, "path to input file")
	part2 := flag.Bool("part2", false, "solve part 2 (timeline count)")
	flag.Parse()
	if flag.NArg() > 0 && flag.Arg(0) != "" {
		inputPath = flag.Arg(0) // allow positional override for compatibility
	}

	grid, err := readLines(inputPath)
	if err != nil {
		log.Fatalf("read input %q: %v", inputPath, err)
	}

	if *part2 {
		fmt.Println(countTimelines(grid))
		return
	}

	fmt.Println(countSplits(grid))
}
